New module: Add Pritunl VPN user module (net_tools/pritunl/)

Lowess opened the pull request:
<!---
Verify first that your issue/request is not already reported on GitHub.
Also test if the latest release, and master branch are affected too.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
pritunl_user.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0 (devel 44fac2eb5a) last updated 2017/05/04 00:58:43 (GMT -700)
```

##### CONFIGURATION
<!---
Mention any settings you have changed/added/removed in ansible.cfg
(or using the ANSIBLE_* environment variables).
-->

##### SUMMARY
<!--- Explain the problem briefly -->

[Pritunl](https://pritunl.com/) is an Enterprise Distributed OpenVPN Server. We have been using this technology for about a year and we felt really handy to automate our VPN user management using Ansible.

As Pritunl offers a [REST API](https://pritunl.com/api), we developed two modules to manage VPN users.

Both modules have been sitting on my laptop for about a year. They have been well tested and they work as expected (respect of idempotency, based on urllib, etc...)

This PR was originally tracked here https://github.com/ansible/ansible/pull/26584
Florian

Lowess commented:
@Akasurde FYI

Andersson007 commented on plugins/modules/net_tools/pritunl/pritunl_user.py:
```diff
@@ -0,0 +1,415 @@
+    user_email: foo@bar.com
+    user_disabled: yes
+
+# Make sure the user Foo is not part of MyOrg anymore
```
```suggestion
```

Andersson007 reviewed (commented):
@Lowess thanks for the PR
1) split the module into two (the other one for getting a list)
2) we need unit tests

felixfontein commented:
@Lowess are you still working on this?

needs_info

Lowess commented:
@felixfontein sorry for the delay here... I am not actively working with Ansible lately but I am willing to wrap up this PR if you guys can point me to the Ansible docs to setup and test this code locally (What ansible version shall I use ? How to simply run tests for this module only). I have been checking multiple documentation resources but they seem a bit outdated.

If you guys have guidance on how to set this up, I will take it.

felixfontein commented:
@Lowess for testing locally, both Ansible 2.9 and ansible-base 2.10 should be fine. CI tests both (and ansible-core devel branch), but in most cases it shouldn't make a difference.

As for tests, we usually prefer integration tests which run in CI (if possible), which for online services seldomly is possible :) So in your case, integration tests that do not run in CI (`unsupported` in `aliases` file), and/or unit tests are a possible alternative.

If you are interested in writing unit tests, the [open_url testing framework from community.internal_test_tools](https://github.com/ansible-collections/community.internal_test_tools/blob/main/tests/unit/utils/open_url_framework.py) can be useful. The community.internal_test_tools collection is available during unit and integration tests. You can find an example (for an inventory plugin) here: https://github.com/ansible-collections/community.hrobot/blob/main/tests/unit/plugins/inventory/test_robot.py

If you're interested in writing `unsupported` integration tests, take a look at one of the existing ones, like https://github.com/ansible-collections/community.general/tree/main/tests/integration/targets/scaleway_ip - these can be run with `ansible-test integration` and are essentially roles.

Lowess commented:
Got it thanks for the pointers, I will try to tackle that next week !

Lowess commented:
@Andersson007 @felixfontein,

I have put some more efforts in this PR and refactored a lot from the first draft you already reviewed:

* `module_utils/net_tools/pritunl/api.py` is now less coupled with Ansible concepts (no more modules passing around, just api calls that can be more easily tested)
* Added `pritunl_user_info` as a module
* Adjusted `pritunl_user` to only do present / absent

I had quite a hard time mocking all the API calls + I am not really used to `unittest` framework (I only know `pytest`). I am fairly new to "mocking everything" so I hope it's not too dirty :)

I Would love to have some insights on the CI build issues I am facing:

* What do I have to do to get `ansible-doc` find the new modules ?

Appart from that seems like CI is happy

Lowess commented:
> I Would love to have some insights on the CI build issues I am facing:
> 
> * What do I have to do to get `ansible-doc` find the new modules ?
> 

Damn I figured out it was due to the missing symlink...

Lowess commented on plugins/modules/net_tools/pritunl/pritunl_user_info.py:
```diff
@@ -0,0 +1,196 @@
+        dict(
+            pritunl_url=dict(required=True, type="str"),
+            pritunl_api_token=dict(required=True, type="str"),
+            pritunl_api_secret=dict(required=True, type="str"),
```
Good catch

Lowess commented on plugins/modules/net_tools/pritunl/pritunl_user_info.py:
```diff
@@ -0,0 +1,196 @@
+        type: str
+        required: true
+        description:
+            - URL and port of the pritunl server on which the API is enabled
```
Done

Lowess commented:
@felixfontein I addressed your PR comments and suggestions about centralizing common args and doc in fragment + minor things you spot out.

felixfontein commented on plugins/modules/net_tools/pritunl/pritunl_user_info.py:
```diff
@@ -0,0 +1,168 @@
+        )
+    ),
+
+    module = AnsibleModule(argument_spec=argument_spec)
```
```suggestion
    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)
```
Info modules must never modify anything, and thus have to support check mode.

felixfontein commented on plugins/modules/net_tools/pritunl/pritunl_user.py:
```diff
@@ -0,0 +1,361 @@
+        default: null
+        description:
+            - Name of the user to create in Pritunl. The I(user_name) is used
+              when the module is used with C(state=list).
```
```suggestion
              when the module is used with I(state=list).
```
For `option=value`, we currently use `I(...)` (https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation).

Lowess replied:
This is actually not valid anymore as this was when I had `pritunl_user_info` part of `pritunl_user`. I will completely remove it

felixfontein commented on plugins/modules/net_tools/pritunl/pritunl_user_info.py:
```diff
@@ -0,0 +1,168 @@
+users:
+    description: List of Pritunl users
+    returned: success
+    type: list
```
```suggestion
    type: list
    elements: dict
```

felixfontein reviewed (commented):
Looks good! I've found some small details, but after that I think it's ready for merging.

Lowess commented:
@felixfontein any idea when this will be merged so that I can tackle PR #804 ?

felixfontein commented:
@Andersson007 do you want to take another look, or should I merge?

Andersson007 commented on plugins/modules/net_tools/pritunl/pritunl_user.py:
```diff
@@ -0,0 +1,360 @@
+
+RETURN = """
+response:
+    description: JSON representation of Pritunl Users
```
```suggestion
    description: JSON representation of Pritunl Users.
```

Andersson007 commented on plugins/modules/net_tools/pritunl/pritunl_user_info.py:
```diff
@@ -0,0 +1,169 @@
+
+RETURN = """
+users:
+    description: List of Pritunl users
```
```suggestion
    description: List of Pritunl users.
```

Andersson007 reviewed (commented):
@Lowess thanks for doing this. Looks impressive.
We forgot to put the dots. In RETURN section it's necessary according to the doc.

Andersson007 commented:
@felixfontein when the dots are added, feel free to merge:)

russoz commented on plugins/module_utils/net_tools/pritunl/api.py:
```diff
@@ -0,0 +1,307 @@
+    )
+
+    if response.getcode() != 200:
+        raise PritunlException("Could not retrive organizations from Pritunl")
```
typo: `retrieve`

Lowess replied:
Good catch my dyslexia hit me again

russoz commented on plugins/module_utils/net_tools/pritunl/api.py:
```diff
@@ -0,0 +1,307 @@
+        validate_certs=dict(required=False, type="bool", default=True),
+    )
+
+
```
hi @Lowess this is not a problem per se, more a suggestion: function above uses `dict()` and function below uses `{}`. For the sake of consistency it might be interesting to have them both use the same construct - whichever one you prefer.

Lowess replied:
@russoz it's just that this util function is used in modules where the `dict()` format is pretty common for `argument_spec` hence why I used `dict()` instead of `{}`

```
argument_spec = pritunl_argument_spec()

argument_spec.update(
        dict(
            organization=dict(required=True, type="str", aliases=["org"]),
            state=dict(
                required=False, choices=["present", "absent"], default="present"
            ),
            user_name=dict(required=True, type="str"),
            user_type=dict(
                required=False, choices=["client", "server"], default="client"
            ),
            user_email=dict(required=False, type="str", default=None),
            user_groups=dict(required=False, type="list", elements="str", default=None),
            user_disabled=dict(required=False, type="bool", default=None),
            user_gravatar=dict(required=False, type="bool", default=None),
        )
    ),
```

With that in mind do you still want me to change to `{}`

russoz replied:
hi @Lowess this is more a matter of personal preferences. I'd prefer them equal, but if you are happy with them as is, no worries, we can live with that. Thanks for the response :-)

russoz commented on plugins/module_utils/net_tools/pritunl/api.py:
```diff
@@ -0,0 +1,307 @@
+                orgs.append(org)
+
+            else:
+                filtered_flag = False
+
+                for filter_key, filter_val in iteritems(filters):
+                    if filter_val != org[filter_key]:
+                        filtered_flag = True
+
+                if not filtered_flag:
+                    orgs.append(org)
```
```suggestion
                if not any(filter_val != org[filter_key] for filter_key, filter_val in iteritems(filters)):
                    orgs.append(org)
```
Go forth and pythonify the world. :grin:

russoz commented on plugins/module_utils/net_tools/pritunl/api.py:
```diff
@@ -0,0 +1,307 @@
+    )
+
+    if response.getcode() != 200:
+        raise PritunlException("Could not retrive users from Pritunl")
```
```suggestion
        raise PritunlException("Could not retrieve users from Pritunl")
```

russoz commented on plugins/modules/net_tools/pritunl/pritunl_user_info.py:
```diff
@@ -0,0 +1,169 @@
+
+    if len(org_obj_list) == 0:
+        module.fail_json(
+            msg="Can not list users from the organization '%s' which does not exist"
```
This is the error message the user will see if the HTTP request in `list_pritunl_organiazations()` returns anything other than a `200` code. This happens because of the `pass` in line 117.

Lowess replied:
Will remove the try catch

russoz commented on plugins/modules/net_tools/pritunl/pritunl_user.py:
```diff
@@ -0,0 +1,360 @@
+            result["response"] = response
+        except PritunlException:
+            module.fail_json(
+                msg="Could not remove user %s from organization %s from Pritunl"
```
Suggestion: instead of putting `try/except` inside each function, raise the exception with the message that should be in the `fail_json()` call, as in:
```python
    raise PritunlException("Could not remove user...")
```
then in `main()` you put the `try/except` around the whole thing, as in:
```python
    try:
        if state == "present":
            add_or_update_pritunl_user(module)
        elif state == "absent":
            remove_pritunl_user(module)
    except PritunlException as e:
        module.fail_json(msg=str(e))
```

russoz commented on tests/unit/plugins/modules/net_tools/pritunl/test_pritunl_user.py:
```diff
@@ -0,0 +1,208 @@
+                    "pritunl_api_token": "token",
+                    "pritunl_api_secret": "secret",
+                    "pritunl_url": "https://pritunl.domain.com",
+                    "organization": "GumGum",
```
Night at the Museum? :rofl:

Lowess replied:
Nop my company name -> https://gumgum.com/ 😛

russoz replied:
Ha, once we have kids and start watching those movies, we see them everywhere :-P

Lowess commented on plugins/module_utils/net_tools/pritunl/api.py:
```diff
@@ -0,0 +1,307 @@
+                users.append(user)
+
+            else:
+                filtered_flag = False
```
@russoz I have also applied the `any` magic here as well

russoz commented on plugins/modules/net_tools/pritunl/pritunl_user.py:
```diff
@@ -0,0 +1,344 @@
+            remove_pritunl_user(module)
+    except PritunlException as e:
+        module.fail_json(msg=str(e))
+
```
It occurred to me later on, it would be good to have a
```
except Exception e:
   module.fail_json("Unknown error: {0}".format(e))
```
to catch anything in there that is *not* a `PritunlException`.

Lowess replied:
Indeed it's a good practice ! will do

russoz commented on plugins/modules/net_tools/pritunl/pritunl_user.py:
```diff
@@ -0,0 +1,344 @@
+            )
+        )
+    except PritunlException:
+        pass
```
This is still not a good way of handling the issue, IMO. The root cause of the problem will be discarded here and the user will receive the "Can not remove user ..." message below. The gap between the error message the user sees and the real problem that occurred will only make it harder to debug.

russoz replied:
Personally I'd remove this try/except here entirely. Let the exception cut through all the layers and the exact error to be show to the user.

felixfontein replied:
Alternatively, use `from ansible.module_utils.six import raise_from` and then `raise_from(NewException(), old_exception)`. If you return the exception in the `fail_json` call with `exception=traceback.format_exc()`, you can see both exceptions (no idea how it looks with Python 2, but with Python 3 it should work well).

Lowess commented on plugins/modules/net_tools/pritunl/pritunl_user.py:
```diff
@@ -0,0 +1,345 @@
+        elif state == "absent":
+            remove_pritunl_user(module)
+    except PritunlException as e:
+        module.fail_json(msg=to_native(e))
```
Made use of `to_native` here as per a previous suggestion from @Andersson007.
And added global `Exception` catching as per @russoz

Lowess commented on plugins/modules/net_tools/pritunl/pritunl_user_info.py:
```diff
@@ -0,0 +1,173 @@
+
+    try:
+        get_pritunl_user(module)
+    except PritunlException as e:
```
Same here I wrapped the call in the same try / catch logic

Lowess commented on plugins/modules/net_tools/pritunl/pritunl_user.py:
```diff
@@ -0,0 +1,352 @@
+            remove_pritunl_user(module)
+    except PritunlException as e:
+        module.fail_json(msg=to_native(e))
+    except (AnsibleExitJson, AnsibleFailJson) as e:
```
Made use of to_native here as per a previous suggestion from @Andersson007.
And added global Exception catching as per @russoz

I had to deal with `AnsibleExitJson` and `AnsibleFailJson` here as I added to global `Exception` catch so that unit tests can still do their job

Lowess replied:
Actually this does not seem doable @russoz according to CI tests. @russoz any idea on how I can do a global exception catch will being compliant with mocking in unittests that uses raise `AnsibleExitJson`

https://docs.ansible.com/ansible/latest/dev_guide/testing_units_modules.html#handling-exit-correctly

felixfontein replied:
You basically can't (in a sane way). The ony way is to avoid a generic catch-everything `except` clause.

Lowess replied:
Thanks @felixfontein I have removed those then...

russoz commented:
LGTM

felixfontein commented:
@Lowess thanks for contributing this!
@Andersson007 @russoz thanks for reviewing!
